snafu-derive/src/lib.rs

```diff
@@ -165,8 +167,18 @@ fn parse_snafu_enum(
                 match attr {
                     SnafuAttribute::Display(d) => display_format = Some(d),
                     SnafuAttribute::Visibility(v) => visibility = Some(v),
-                    SnafuAttribute::Source(..) => { /* Report this isn't valid here? */ }
-                    SnafuAttribute::Backtrace(..) => { /* Report this isn't valid here? */ }
+                    SnafuAttribute::Source(..) => {
+                        errors.push(syn::Error::new(
+                            name.span(),
+                            "'source' attribute is only valid on individual fields; attempted to place on this variant:",
+                        ));
```

Comment:
I wonder if there’s anything we can do about the “programmer overhead” of creating these, as well as the duplicated phrases. Potentially, this could look like:

```rust
errors.only_on_field(name, “source”)
errors.only_on_field(name, “backtrace”)
```

As a minimum, this could be used to hide the inner `Vec` and construction of `Error`:

```rust
impl MyErrorCollectionName {
    fn add(&mut self, span: impl Spanned, text: impl Into<String>) { /* … */ }

    fn finish(self) -> Result<(), Vec<…>> { /* … */ }
}
```

Reply:
I lean toward the latter because I think having a custom message can be helpful for clarifying what the user actually did that wasn't quite right.  I'll give that a shot.

Reply:
> I lean toward the latter

I don’t think they have to be either-or:

```rust
impl MyErrorCollectionName {
    fn add_only_on_field(&mut self, span: impl Spanned, field: &str) {
        self.add(span, format!(“This is repeated text for field {}, do this other thing instead”));
    }

    fn add(&mut self, span: impl Spanned, text: impl Into<String>) { /* … */ }
}
```

Reply:
Ahh, I see what you mean.  "field" works well for handling duplication, but we might also want something like "element type" for misuse?  I think I'd typically want to say something like "this doesn't apply to [field, variant, enum, struct], it's only valid on [field, variant, enum, struct]".  I guess that could just be handled by the custom message, but it's even more common than repetition, so maybe it's worth a try simplifying it too.

Reply:
Yeah, there's lots of ways this could go. My suggestion would be to start with adding the plain `add` method and then see what repetition remains. That might then highlight new methods to be created. How those look could be as simple as taking in a `&str`, up to taking in a slice of items, or as far as some custom type(s). I think it's worth staying practical with respect to the amount of code we are saving, so the latter option seems unlikely :-) 

As a side note, you can't actually use the `impl ...` syntax I listed because we support Rust 1.18, but hopefully the idea is communicated.

Reply:
I started with `add`, moved to a couple custom methods that covered 90% of the cases, and eventually found that a couple custom types were simpler.  While it's slightly more verbose, I think it's rather clear - hope you like it as much as I do :)

Reply:
> eventually found that a couple custom types were simpler

I don't disagree, but would you mind expanding on bit on how you found the methods lacking?

Reply:
Sure - in particular it was `SyntaxErrors.only_valid_on`, which became the `OnlyValidOn` struct.  It had 3 arguments for descriptive strings, and 1 for the span.  The 3 string arguments were too easy for a developer to mix up, in my opinion -- you always had to check the docs.  With the struct, where you specify its fields at creation, it becomes obvious, and I think it reads very nicely.

(I think having those 3 fields are important to clarify the different types of errors, and clearly point out to the user what they've done wrong.)

Reply:
> The 3 string arguments were too easy for a developer to mix up, in my opinion

I had a hunch this was it. For future reference, one thing I've seen Niko Matsakis do is pseudo-named-arguments:

```rust
struct Attribute(&'static str);
struct ValidOn(&'static str);
struct NotOn(&'static str);

fn only_valid_on(attribute: Attribute, valid_on: ValidOn, not_on: NotOn) {
    println!(
        "'{}' attribute is only valid on {}; attempted to place on this {}:",
        attribute.0, valid_on.0, not_on.0
    )
}

fn main() {
    only_valid_on(Attribute("a"), ValidOn("b"), NotOn("c"));
}
```

Reply:
I'm happy to go with whichever style you prefer :)